design/one-pager-moving-packages.md

```diff
@@ -0,0 +1,459 @@
+Assume you install a Configuration that depends on
+`xpkg.upbound.io/crossplane-contrib/provider-nop`, which is later renamed to
+`xpkg.upbound.io/crossplane/provider-nop`. You then install an updated version
+of the Configuration, which now depends on the renamed Provider. These providers
+are the identical, but Crossplane considers them to be different because the OCI
+reference changed.
+
+The resolver controller creates a new Provider named `crossplane-provider-nop`,
+which conflicts with the existing Provider named
+`crossplane-contrib-provider-nop`. The conflict occurs because both Providers
+want to have an active ProviderRevision that controls their CRDs, but only one
+ProviderRevision may be the CRDs' [controller reference][controller-ref].
+
+Replicating a dependency breaks because dependency references are hardcoded in
+package metadata, and are fully qualified - they specifically include the OCI
+registry to pull from. You can replicate a package from `xpkg.upbound.io` to
+`pkg.internal.acme.co`, but its dependencies are still hard coded to
+`xpkg.upbound.io`.
+
+## Goals
+
+The goal of this proposal is to ensure dependencies don't break when you:
+
+1. Move a package to a different registry, organization, or repository
+2. Replicate a package to a different, potentially private, registry
+
+Handling private registry authentication for dependencies isn't in scope for
+this design. It's covered by the [package `ImageConfig` proposal][imageconfig].
+
+## Proposal
+
+I propose we do two things:
+
+1. Allow packages to specify that they replace another package
+2. Allow packages to specify partially qualified dependencies
+
+These proposed changes correspond to the goals above. The first addresses
+moving dependencies, while the second addresses replicating dependencies.
+
+### Moving a Package
+
+I propose we support moving a package by adding a `replaces` stanza to package
+metadata.
+
+For example, assume the maintainers of `provider-family-aws` move it from
+`xpkg.upbound.io/upbound/provider-family-aws` to
+`xpkg.upbound.io/crossplane-contrib/provider-family-aws`. The maintainers would
+add the following `replaces` stanza to the moved provider:
+
+```yaml
+apiVersion: meta.pkg.crossplane.io/v1
+kind: Provider
+metadata:
+  name: provider-family-aws
+spec:
+  replaces:
+  - xpkg.upbound.io/upbound/provider-family-aws
+```
+
+This tells Crossplane that the `crossplane-contrib/provider-family-aws` package
+replaces the `upbound/provider-family-aws` package.
+
+When the revision controller encounters a package with a `replaces` stanza,
+it'll deactivate the replaced package(s) before installing the replacement
+package.
+
+After Crossplane deactivates the replaced package, it can install the
+replacement package without a conflict. A deactivated package [owns][owner-ref]
+but doesn't [control][controller-ref] its payload (e.g. a Provider's MR CRDs).
+This eliminates the controller reference conflict.
```

Comment:
Probably a little stretch, but still sharing in case it triggers other thoughts/edge cases.

What happens in the following scenario:

I have the below packages:

```
- registry1.com/acme/configuration-foo:v0.1.0
  -  registry1.com/acme/provider-nop >= v0.1.0
```

```
- registry1.com/acme/configuration-bar:v0.1.0
  -  registry1.com/acme/provider-nop <= v0.2.0
```

and `registry1.com/acme/provider-nop:v0.2.0` installed by package manager.

I decided to move `registry1.com/acme/provider-nop` to `registry1.com/acme-providers/provider-nop`, added the replaces stanza. Updated one of the configuration to use the new one, but the other one not updated yet:

```
- registry1.com/acme/configuration-foo:v0.1.0
  -  registry1.com/acme-providers/provider-nop >= v0.1.0
```

Package manager will see the replaces stanza and deactivate the `registry1.com/acme/provider-nop` and install the new one, say, `registry1.com/acme-providers/provider-nop:v0.2.1`. At this state, `registry1.com/acme/configuration-bar` should be unhappy because its dependency no longer satisfied, since it was replaces with a new version which it doesn't like.

Reply:
Right, in this scenario the replaced package will:

* Exist in a deactivated form
* Won't appear in the lock

So because a configuration still depends on it, the resolver controller is going to see that it's not in the lock and try to install it. When it does, it'll "succeed" because it'll try create the provider and silently ignore the error it hits because the (deactivated) provider already exists.

Meanwhile, the configuration that depends on the replaced package will become unhealthy because the revision controller will notice its dependencies are no longer satisfied.

One important side effect of this is that it's going to jam up the revision controller, which only processes the first implied (i.e. missing) dependency on each reconcile. It'll get stuck in a loop of noticing this dependency is missing, and creating a provider that already exists. It'll never move on to the next dependency.

So one thing we should probably do is have the resolver controller reconcile a random missing dependency, not the first every time. That way it can at least make progress in this situation.

We could also consider adding `replaces` to the lock, e.g.:

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: Lock
metadata:
  name: lock
packages:
- name: upbound-configuration-getting-started-26746e8e7abb
  type: Configuration
  source: xpkg.upbound.io/upbound/configuration-getting-started
  version: v0.2.0
  dependencies:
  - constraints: v0.2.1
    package: xpkg.upbound.io/crossplane-contrib/provider-nop
    type: Provider
- name: crossplane-contrib-provider-nop-ecc25c121431
  type: Provider
  source: xpkg.upbound.io/crossplane-contrib/provider-nop
  version: v0.2.1
  replaces:
  - xpkg.upbound.io/upbound/provider-nop
```

Then we could update the DAG logic to consider a package to be installed if either:

* It appears as a top level `source` in the packages array
* It appears in the `replaces` array of any top level entry in the packages array

Reply:
In future we might also reconsider removing packages from the lock when all of their revisions are deactivated.

A deactivated package is in kind of a weird middle ground state:

* It's still technically installed
* Its payload (CRDs, XRDs, etc) still exists (because they have an owner ref to the package, not just revision)

The question is whether we should consider a dependency to be satisfied by a package that's installed, was active, but is now inactive. If we consider the dependency to be on the _types_ (CRDs, XRDs, Compositions etc) those are all still there...

Reply:
Kinda related: https://github.com/crossplane/crossplane/issues/6018